### PR TITLE
Log item self-expansion

### LINKED ISSUE #8527: [Bug]: Cannot merge folders with wildcards
### Issue Description

Trying to use wildcards to merge the contents of two folders looks like it should be possible, but doesn't work due to bugs.

### Steps to Reproduce

I have a vcxproj and two folders of resource files that I want to merge in the output. The content of the folders might be generated so I can't enumerate them in the project file, I need to use wildcards. I know wildcards are not fully supported in vcxproj so I follow the advice of [the docs](https://learn.microsoft.com/en-us/cpp/build/reference/vcxproj-files-and-wildcards) and put them in a Target:

```
<Target Name="SomeTarget" BeforeTargets="ClCompile">
   <ItemGroup>
      <Content Include="files1\**" CopyToOutputDirectory="PreserveNewest" TargetPath="files\@(RecursiveDir)%(Filename)%(Extension)"/>
      <Content Include="files2\**" CopyToOutputDirectory="PreserveNewest" TargetPath="files\@(RecursiveDir)%(Filename)%(Extension)"/>
   </ItemGroup>
</Target>
```

My project zipped:

[CppSandbox.zip](https://github.com/dotnet/msbuild/files/10883572/CppSandbox.zip)

### Expected Behavior

The contents of the two folders should be merged at buildtime.

### Actual Behavior

The project fails to build with the following error:
> error MSB3024: Could not copy the file "D:\Dev\Sandbox\CppSandbox\files1\a.txt" to the destination file "D:\Dev\Sandbox\CppSandbox\Debug\files\", because the destination is a folder instead of a file. To copy the source file into a folder, consider using the DestinationFolder parameter instead of DestinationFiles.

Contrary to the docs, if I omit the Target and use a direct ItemGroup, it works. Another weird behaviour occurs if we consider one of the folders to be generated at buildtime, ergo not present at project evaluation time. So let's just put that in a Target:
```
<ItemGroup>
	<Content Include="files1\**" CopyToOutputDirectory="PreserveNewest" TargetPath="files\@(RecursiveDir)%(Filename)%(Extension)"/>			
</ItemGroup>
<Target Name="SomeTarget" BeforeTargets="ClCompile">
	<ItemGroup>
		<Content Include="files2\**" CopyToOutputDirectory="PreserveNewest" TargetPath="files\@(RecursiveDir)%(Filename)%(Extension)"/>
	</ItemGroup>
</Target>
```
Now MSBuild gets confused and starts overwriting the files with each other:
>  Copying file from "D:\Dev\Sandbox\CppSandbox\files1\c.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\c.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files1\a.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\a.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files1\b.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\b.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files2\d.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\b.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files2\d.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\c.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files2\d.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\a.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files2\e.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\b.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files2\e.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\c.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files2\e.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\a.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files2\f.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\b.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files2\f.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\c.txt".
  Copying file from "D:\Dev\Sandbox\CppSandbox\files2\f.txt" to "D:\Dev\Sandbox\CppSandbox\x64\Debug\files\a.txt". 

### Analysis

_No response_

### Versions & Configurations

Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.
16.11.2.50704

or

MSBuild version 17.5.0+6f08c67f3 for .NET Framework
17.5.0.10706

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes #8527

### Context
Self referencing item metadata in an item definition within the target leads to possible unintended expansion (and cross-applying of pre-existing item instances).

This behavior is not a bug - it's the implication of the target batching feature - but might be confusing. So detection and warning is added here

### Changes Made
A self reference (qualified or unqualified) of metadata within the item defeiniton (which is within a target) is detected and high importance message is issued.

### Testing
Tests added for a warning case and for a non-warning case (self-referencing metadata outside of target context)

### Doc
https://github.com/MicrosoftDocs/visualstudio-docs-pr/pull/11034

### Possible impact
It's unfortunately hard to obtain data on prevalence of this pattern due to limitations of available code searches (github search doesn't support '@' sign and no escaping option; SourceGraph, grep.app doesn't support regex lookaheads). So I cannot quantify (I'll try again later on).
But there are some sparse evidence of usage:

- https://github.com/JeremyAnsel/JeremyAnsel.HLSL.Targets/blob/master/JeremyAnsel.HLSL.Targets/JeremyAnsel.HLSL.Targets/JeremyAnsel.HLSL.Targets.csproj#L49
- https://github.com/neuecc/MessagePack-CSharp/blob/master/src/MessagePack.MSBuild.Tasks/MessagePack.MSBuild.Tasks.csproj#L35
- https://github.com/dotnet/aspnetcore/blob/main/eng/AfterSigning.targets#L22

all of those seem to be doing something else then intended (the variable part of path is empty, so it gets just the base directory), but they do not break.

**tl;dr;:** ~~unless we are able to better quantify, we might resort to demote the warning to a message~~ Demoted to Message severity

### CODE REVIEW

File: src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupIntrinsicTask.cs [hunk 1]
```diff
@@ -184,7 +184,20 @@ private void ExecuteAdd(ProjectItemGroupTaskItemInstance child, ItemBucket bucke
 
                 if (condition)
                 {
-                    string evaluatedValue = bucket.Expander.ExpandIntoStringLeaveEscaped(metadataInstance.Value, ExpanderOptions.ExpandAll, metadataInstance.Location, loggingContext);
+                    ExpanderOptions expanderOptions = ExpanderOptions.ExpandAll;
+                    ElementLocation location = metadataInstance.Location;
+                    if (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_6) &&
```
Comment (Forgind) [inline src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupIntrinsicTask.cs:189]: Since this never logs a warning anymore, we might want to rename the ExpanderOptions, and since adding a message isn't a breaking change, we probably don't need this anymore.

(And if you think we still need the changewave, please it to the doc!)

File: src/Build.UnitTests/BackEnd/MSBuild_Tests.cs [hunk 2]
```diff
@@ -772,6 +772,138 @@ public void ItemsIncludeExcludePathsCombinations()
...
+        /// <summary>
+        /// Referring to an item outside of target leads to 'naturally expected' reference to the item being processed.
+        ///  No expansion occurs.
+        /// </summary>
+        [Fact]
+        public void ItemsRecursionOutsideTarget()
+        {
+            string projectFile = null;
+
+            try
```
Comment (Forgind) [inline src/Build.UnitTests/BackEnd/MSBuild_Tests.cs:784]: You can probably make these a little cleaner with TestEnvironment. It does the cleanup for you automatically when it's disposed.

File: src/Build.UnitTests/BackEnd/MSBuild_Tests.cs [hunk 3]
```diff
@@ -772,6 +772,138 @@ public void ItemsIncludeExcludePathsCombinations()
...
+                ObjectModelHelpers.BuildTempProjectFileExpectSuccess(projectFile, logger);
+
+                Console.WriteLine(logger.FullLog);
+
+                logger.AssertLogDoesntContain("iin1=[a/b.foo;c/d.foo;g/h.foo]");
+                logger.AssertLogDoesntContain("iin1-target-paths=[b.foo;d.foo;h.foo]");
+                logger.AssertLogContains("iin1=[a/b.foo;c/d.foo;g/h.foo;g/h.foo]");
+                logger.AssertLogContains("iin1-target-paths=[;b.foo;b.foo;d.foo]");
+
+                logger.AssertLogContains("MSB4120: Item 'iin1' definition within target is referencing self via metadata 'Extension' (qualified or unqualified). This can lead to unintended expansion and cross-applying of pre-existing items");
```
Comment (Forgind) [inline src/Build.UnitTests/BackEnd/MSBuild_Tests.cs:849]: This is English-specific so will fail on non-English systems. Can you get it from the resource directly?

**Review by Forgind — COMMENTED** [review]
I think this should work. I wasn't actually aware of how confusing the self-referential behavior is before you brought it up, so I appreciate it!

For the cases when people are using this in the wild, were they using it intentionally, or were they broken and just confused?

File: src/Build/Resources/Strings.resx [hunk 4]
```diff
@@ -1983,4 +1983,8 @@ Utilization:          {0} Average Utilization: {1:###.0}</value>
   <data name="NodeReused" xml:space="preserve">
     <value>Reusing node {0} (PID: {1}).</value>
   </data>
-</root>
\ No newline at end of file
+  <data name="ItemReferencingSelfInTarget" xml:space="preserve">
+    <value>MSB4120: Item '{0}' definition within target is referencing self via metadata '{1}' (qualified or unqualified). This can lead to unintended expansion and cross-applying of pre-existing items. More info: https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching#item-batching-on-self-referencing-metadata</value>
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/Resources/Strings.resx:1987]: Unfortunately HelpLink is currently only supported by Warnings and Errors (and it actually doesn't seem to be populated by any of our code other than tests). So it would likely got lost - I prefer keeping it in the message

File: src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupIntrinsicTask.cs (hunk 1 above)
Comment (JanKrivanek) [reply src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupIntrinsicTask.cs:189]: Thanks for reminding to document!

I'd still prefer to be defensive - as the changewave guards not only the logging, but as well the heavy-lifting to detect the situation

File: src/Build/Definition/ProjectItemDefinition.cs [hunk 5]
```diff
@@ -26,7 +26,7 @@ namespace Microsoft.Build.Evaluation
     /// ProjectMetadataElement, and these can be added, removed, and modified.
     /// </remarks>
     [DebuggerDisplay("{_itemType} #Metadata={MetadataCount}")]
-    public class ProjectItemDefinition : IKeyed, IMetadataTable, IItemDefinition<ProjectMetadata>, IProjectMetadataParent
+    public class ProjectItemDefinition : IKeyed, IMetadataTable, IItemDefinition<ProjectMetadata>, IProjectMetadataParent, IItemMetadata
```
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/Definition/ProjectItemDefinition.cs]: Sorry if I misunderstood here. I renamed the interface for clarity (`IItemMetadata` -> `IItemTypeDefinition`) - hope this is adressing the concern.

File: src/Build/Evaluation/IItemTypeDefinition.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable disable
```
Comment (Forgind) [inline src/Build/Evaluation/IItemTypeDefinition.cs:4]: nit:
Do you need to disable nullable here?

**Review by Forgind — APPROVED** [review]
LGTM, thanks!

File: src/Build/Evaluation/IItemTypeDefinition.cs (hunk 6 above)
Comment (JanKrivanek) [reply src/Build/Evaluation/IItemTypeDefinition.cs]: Removed
